### DISCUSSION TITLE
Using graphql-constraint-directive

### DISCUSSION THREAD

**Discussion by leo-petrucci** [prisma/prisma#3715] [Q&A] [answered]
I've been looking for a way to add constraints to the arguments of my mutations to enforce things like minimum length of strings etc.

I've found [graphql-constraint-directive](https://github.com/confuser/graphql-constraint-directive), which looks exactly like what I need, unfortunately it requires editing my `schema.graphql` file which is generated by Nexus.

I really like Nexus so I'd rather not ditch it, but constraints are quite important so I'm finding myself kind of stuck.

Any suggestions on how to implement this while working with Prisma and Nexus?

**Comment by leo-petrucci** [comment, marked as answer]
At the moment I've had to resort to adding:
```ts
        if (title && title.length < 20) {
          return new Error('Title must be at least 20 characters long.')
        }
```

To my mutations, which also requires me to change the type of the return of `resolve` to `: Promise<any>`. Which doesn't seem too elegant.

**Reply by ryands17** [reply]
Hey @creativiii 👋 

I think one option would be to use something like [yup](https://github.com/jquense/yup). An example resolver would look like this:

```ts
const validationSchema = {
  signup: yup.object().shape({
    email: yup.string().email(),
    password: yup.string().min(8),
  }),
}

t.field('login', {
      type: 'AuthPayload',
      args: {
        email: schema.stringArg({ required: true }),
        password: schema.stringArg({ required: true }),
      },
      resolve: async (_root, args, ctx) => {
        await validationSchema.signup.validate(args)

        let user = null
        try {
          user = await ctx.db.user.findOne({
            where: {
              email: args.email,
            },
          })
        } catch (e) {
          throw Error('Invalid username or password!')
        }

        if (!user) throw Error('Invalid username or password!')

        const passwordValid = await compare(args.password, user.password)
        if (!passwordValid) throw Error('Invalid username or password!')

        const accessToken = generateAccessToken(user.id)
        return {
          accessToken,
          user,
        }
      },
})
```

**Reply by leo-petrucci** [reply]
This is a great solution!

Ended up doing something like this to return custom errors:

```
const validationSchema = {
  editTitle: object().shape({
    title: string().min(10, 'Title must be at least 10 characters long.'),
  }),
}

try {
  await validationSchema.editTitle.validate({ title })
} catch (e) {
  return new Error(e.errors[0])
}
````

Thank you for the help!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
